### PR TITLE
Fixing trailing line issue

### PR DESCRIPTION
**Pull request by jaimehisao**
There's an issue that caused lines to be added to the last chunk when multiple lines were deleted, this has been addressed with this PR, and tests have been added to make sure of it.

### CODE REVIEW

**Comment by jaimehisao** [comment]
Worth mentioning git issues regarding multiple unrelated commits being added has been fixed, we are now tracking master from google repo instead of my fork.

File: src/patch_parser.py [hunk 1]
```diff
@@ -82,6 +85,15 @@ def set_previous_line(self, item):
     def get_previous_line(self):
         return self._previous_item
 
+    def get_number_of_deleted_lines(self):
```
Comment (dlatypov) [inline src/patch_parser.py]: Suggestion:

Let's consider adding something like

```python
import dataclasses

@dataclasses.dataclass
class HunkParserState:
  gerrit_orig_line: int
  gerrit_new_line: int
  deleted_lines: int
```

Then we can pass that around the parse hunk functions instead of taking multiple arguments and returning updated versions of all of them.

And that gives us a place to store deleted_lines().
It feels like we're using `InputSource` as a handy place to store this additional state (`deleted_lines`) bc the Tuple was getting unwieldy.
But it doesn't really feel like it belongs, given the more generic nature of InputSource.

File: src/patch_parser.py [hunk 2]
```diff
@@ -423,21 +441,33 @@ def _parse_patch_file_unchanged_chunk(
         lines.consume()
         gerrit_orig_line += 1
         gerrit_new_line += 1
-    return (gerrit_orig_line,
+        #logging.warning('Unchanged start: ' + str(in_start))
```
Comment (dlatypov) [inline src/patch_parser.py]: we might want these to keep debugging while refactoring, but reminder to go back and delete these before we merge

File: src/patch_parser.py [hunk 3]
```diff
@@ -423,21 +441,33 @@ def _parse_patch_file_unchanged_chunk(
...
+        #logging.warning('Unchanged line: ' + lines[0])
+        #logging.warning('Unchanged lines - 1: ' + str(lines.line_number() - 1))
+    if lines.get_number_of_deleted_lines() == 0:
```
Comment (dlatypov) [inline src/patch_parser.py]: Aren't these branches effectively the same?

i.e. `gerrit_new_line - lines.line_number() == gerrit_new_line - lines.get_number_of_deleted_lines() - lines.line_number()` when `get_number_of_deleted_lines() == 0`.

Perhaps just adding a new local variable would work better

```
offset = gerrit_new_line - lines.get_number_of_deleted_lines() - lines.line_number()
return ....
```

File: src/patch_parser.py [hunk 4]
```diff
@@ -473,14 +504,14 @@ def _parse_patch_file_removed_chunk(
 
 def _parse_patch_file_chunk(lines: InputSource,
                             gerrit_orig_line: int,
-                            gerrit_new_line: int) -> Tuple[int, int, PatchFileChunkLineMap]:
+                            gerrit_new_line: int, chunks: PatchFileChunkLineMap) -> Tuple[int, int, PatchFileChunkLineMap]:
```
Comment (dlatypov) [inline src/patch_parser.py]: we don't seem to be using the newly added chunks parameter here or in `__parse_patch_file_added_chunk()`.
I assume it's a leftover from before when you were considering inspecting the previous hunk to find `deleted_lines()`?

File: src/patch_parser.py [hunk 5]
```diff
@@ -352,6 +364,12 @@ def map(self, raw_line: int) -> Tuple[str, int]:
             raise IndexError(
                 'Expected ' + str(self.in_range[0]) + ' <= ' + str(raw_line) + ' <= ' + str(self.in_range[1]))
 
+    def get_range(self) -> Tuple[int, int]:
+        return self.in_range
+
+    def get_side(self) -> str:
+        return self.side
```
Comment (dlatypov) [inline src/patch_parser.py]: these new functions seem to be unused now.
I assume this is from a previous attempt at inspecting the chunks map to find `deleted_lines` instead of storing it in InputSource.

So I think we can drop these for now.

File: src/patch_parser_test.py [hunk 6]
```diff
@@ -136,6 +136,43 @@ def test_one_deleted_line(self):
...
+                     line 10  (then 6) # this is line 19 in raw_patch
+
+                    base-commit: 235360eb7cd778d7264c5e57358a3d144936b862
+                    --
+                            '''.strip()
+        line_map = patch_parser._parse_git_patch(raw_patch)
+        self.assertEqual(line_map.map(11), ('file', 2), msg=repr(line_map))  # line 2 (11), control assert
+        self.assertEqual(line_map.map(12), ('file', 3), msg=repr(line_map))  # line 3 (12), control assert
+        self.assertEqual(line_map.map(13), ('fileb', 4), msg=repr(line_map))  # Lines 13 - 16 are deleted and should
+        self.assertEqual(line_map.map(14), ('fileb', 5), msg=repr(line_map))  # show on 'fileb' with their corresponding
```
Comment (dlatypov) [inline src/patch_parser_test.py]: optional: you seem to have line-wrapped the comments, but this one word sticks out.
I think we can just move it into the comment below and everything should fit.

**Review by dlatypov — CHANGES_REQUESTED** [review]

File: src/patch_parser.py [hunk 7]
```diff
@@ -82,6 +85,15 @@ def set_previous_line(self, item):
     def get_previous_line(self):
         return self._previous_item
 
+    def get_number_of_deleted_lines(self):
+        return self._deleted_lines
+
+    def add_deleted_line(self):
+        self._deleted_lines = self._deleted_lines + 1
```
Comment (bjh83) [inline src/patch_parser.py]: I agree with what Daniel said above, so you might move this out or get rid of it entirely. Nevertheless, you could make this:

```
self._deleted_lines += 1
```

File: src/patch_parser.py (hunk 5 above)
Comment (jaimehisao) [reply src/patch_parser.py]: That's right, they are un-used. Let me drop them as well as the other un-used items.

**Review by bjh83 — COMMENTED** [review]
Other than Daniel's comments, LGTM.

File: src/patch_parser.py (hunk 1 above)
Comment (jaimehisao) [reply src/patch_parser.py]: Working on this...

File: src/patch_parser.py (hunk 3 above)
Comment (jaimehisao) [reply src/patch_parser.py]: This is very true! Good catch, I went ahead and changed it.

File: src/patch_parser.py [hunk 8]
```diff
@@ -415,34 +433,41 @@ def _does_match_end_of_super_chunk(lines: InputSource) -> bool:
...
     logging.info('First char - 1: %c', lines[0][0])
     while lines[0] and lines[0][0] == '+':
-        previous = lines.get_previous_line()
         previous = lines.get_previous_line()
         lines.consume()
         gerrit_orig_line += 1
         gerrit_new_line += 1
-        if previous[0] == '-': # maybe add check to see if it is actually a modified one though I think this situation only applies when it actually is modified
+        if previous[
+            0] == '-':  # maybe add check to see if it is actually a modified one though I think this situation only applies when it actually is modified
```
Comment (bjh83) [inline src/patch_parser.py]: Nit: Can you make this a TODO comment and make it a wrapped line comment? This line is pretty long.

File: src/patch_parser.py [hunk 9]
```diff
@@ -11,7 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
+import dataclasses
```
Comment (dlatypov) [inline src/patch_parser.py:15]: very minor nit: can we add a \n before this so there's a separation between license comment and the code?

File: src/patch_parser.py [hunk 10]
```diff
@@ -512,7 +541,7 @@ def _parse_patch_file_super_chunk(lines: InputSource) -> List[PatchFileChunkLine
          gerrit_new_line,
          chunk) = _parse_patch_file_chunk(lines,
                                           gerrit_orig_line,
-                                          gerrit_new_line)
+                                          gerrit_new_line, parser_state)
```
Comment (bjh83) [inline src/patch_parser.py]: Can you put this on a new line to match the rest of the parameters here?

File: src/patch_parser.py [hunk 11]
```diff
@@ -156,6 +156,24 @@ def diff_best_match(self, string: List[str]) -> List[str]:
         return node.diff_best_match(string)
 
 
+@dataclasses.dataclass
+class HunkParserState:
+    _number_of_edits: int
+    _deleted_lines: int
+
+    def get_number_of_deleted_lines(self):
```
Comment (dlatypov) [inline src/patch_parser.py]: nit: I'd drop the getters and setters here since we're using a dataclass.
We can have the callers interact with the fields directly, or use a helper where it makes sense (e.g. if we move gerrit_{orig,new}_line in here, then we could have a helper to increment them all)

File: src/patch_parser.py [hunk 12]
```diff
@@ -156,6 +156,24 @@ def diff_best_match(self, string: List[str]) -> List[str]:
         return node.diff_best_match(string)
 
 
+@dataclasses.dataclass
+class HunkParserState:
+    _number_of_edits: int
```
Comment (dlatypov) [inline src/patch_parser.py]: nit: this field seems unused, let's drop it

File: src/patch_parser.py [hunk 13]
```diff
@@ -415,34 +433,41 @@ def _does_match_end_of_super_chunk(lines: InputSource) -> bool:
...
     in_start = lines.line_number()
     while (not _does_match_end_of_super_chunk(lines)) and (
             (not lines[0]) or (lines[0] and lines[0][0] != '+' and lines[0][0] != '-')):
         logging.info('dropping line: %s', lines[0])
         lines.consume()
         gerrit_orig_line += 1
         gerrit_new_line += 1
+        # logging.warning('Unchanged start: ' + str(in_start))
+        # logging.warning('Unchanged line: ' + lines[0])
+        # logging.warning('Unchanged lines - 1: ' + str(lines.line_number() - 1))
```
Comment (bjh83) [inline src/patch_parser.py]: As I think Daniel mentioned, please drop these comments when you are done or promote them to debug log statements.

File: src/patch_parser.py [hunk 14]
```diff
@@ -156,6 +156,24 @@ def diff_best_match(self, string: List[str]) -> List[str]:
...
+    def get_number_of_deleted_lines(self):
+        return self._deleted_lines
+
+    def add_deleted_line(self):
+        self._deleted_lines += 1
+
+    def reset_deleted_lines(self):
+        self._deleted_lines = 0
+
+    def __init__(self):
```
Comment (dlatypov) [inline src/patch_parser.py]: we don't need this, I think.

But we should be able to also do
```python
deleted_lines: int = 0 
```
up top to be more explicit instead of needing an `__init__()`

File: src/patch_parser.py [hunk 15]
```diff
@@ -415,34 +433,41 @@ def _does_match_end_of_super_chunk(lines: InputSource) -> bool:
 def _parse_patch_file_unchanged_chunk(
         lines: InputSource,
         gerrit_orig_line: int,
-        gerrit_new_line: int) -> Tuple[int, int, PatchFileChunkLineMap]:
+        gerrit_new_line: int,
```
Comment (dlatypov) [inline src/patch_parser.py]: I was originally thinking we could move these gerrit_orig/new_line fields into HunkParserState.
Wdyt?

File: src/patch_parser.py (hunk 8 above)
Comment (dlatypov) [inline src/patch_parser.py]: the added line wrapping here seems weird.
Perhaps let's revert this hunk?

File: src/patch_parser.py [hunk 16]
```diff
@@ -458,9 +483,11 @@ def _parse_patch_file_added_chunk(
 def _parse_patch_file_removed_chunk(
         lines: InputSource,
         gerrit_orig_line: int,
-        gerrit_new_line: int) -> Tuple[int, int, PatchFileChunkLineMap]:
+        gerrit_new_line: int,
+        parser_state: HunkParserState) -> Tuple[int, int, PatchFileChunkLineMap]:
     in_start = lines.line_number()
     while lines[0] and lines[0][0] == '-':
+        parser_state.add_deleted_line()
```
Comment (dlatypov) [inline src/patch_parser.py]: nit: I'd slightly prefer the `lines.consume()` comes first and this integer increment goes down with the rest (`gerrit_orig_line += 1`)

**Review by dlatypov — CHANGES_REQUESTED** [review]

File: src/patch_parser.py (hunk 8 above)
Comment (jaimehisao) [reply src/patch_parser.py]: huh I hadn't noticed that!

File: src/patch_parser.py (hunk 11 above)
Comment (jaimehisao) [reply src/patch_parser.py]: I see, to do everything in one go basically?

File: src/patch_parser.py (hunk 15 above)
Comment (jaimehisao) [reply src/patch_parser.py]: It's fine, I just left them there because of the original design, but I can move them over.

File: src/patch_parser.py (hunk 11 above)
Comment (dlatypov) [reply src/patch_parser.py]: re moving new fields in: yeah, I was thinking it'd be fine to do that in one go.
But if you'd prefer, we could do this separately.

One option could be to do this in two PRs, and another is to clean up the commits a bit so we merge this PR w/o squashing.
I.e. squash all the current commits, and then have a second commit that moves the remaining fields into `HunkParserState`

File: src/patch_parser.py [hunk 17]
```diff
@@ -414,105 +422,97 @@ def _does_match_end_of_super_chunk(lines: InputSource) -> bool:
...
+                                 side='', offset=offset)
 
 
 def _parse_patch_file_added_chunk(
         lines: InputSource,
-        gerrit_orig_line: int,
-        gerrit_new_line: int) -> Tuple[int, int, PatchFileChunkLineMap]:
+        parser_state: HunkParserState) -> PatchFileChunkLineMap:
     in_start = lines.line_number()
+    previous_number_of_removed_lines = parser_state.deleted_lines
```
Comment (dlatypov) [inline src/patch_parser.py]: this var is currently unused, did you want to use it to compute offset?
i.e.
offset=(parser_state.gerrit_new_line - lines.line_number() - previous_deleted_lines
?

or that can be left for another commit and we can drop this var for now.

File: src/patch_parser.py [hunk 18]
```diff
@@ -414,105 +422,96 @@ def _does_match_end_of_super_chunk(lines: InputSource) -> bool:
...
-                            gerrit_new_line: int) -> Tuple[int, int, PatchFileChunkLineMap]:
+                            parser_state: HunkParserState) -> PatchFileChunkLineMap:
     line = lines[0]
     start_line_len = len(lines)
     if _does_match_end_of_super_chunk(lines):
         raise ValueError('Unexpected line: ' + line)
     elif line and line[0] == '+':
         logging.info('First char - 0: %c', line[0])
-        ret_val = _parse_patch_file_added_chunk(lines, gerrit_orig_line, gerrit_new_line)
+        ret_val = _parse_patch_file_added_chunk(lines, parser_state)
```
Comment (dlatypov) [inline src/patch_parser.py]: nit: while we're here, can we update ret_val to something better?

It was called that since it was a tuple and didn't have a good name.
perhaps `chunk_map` ?

File: src/patch_parser.py [hunk 19]
```diff
@@ -414,105 +422,96 @@ def _does_match_end_of_super_chunk(lines: InputSource) -> bool:
...
-        gerrit_orig_line += 1
-        gerrit_new_line += 1
-    return (gerrit_orig_line,
-            gerrit_new_line,
-            PatchFileChunkLineMap(in_range=(in_start, lines.line_number() - 1),
-                                  side='',
-                                  offset=(gerrit_new_line - lines.line_number())))
+        parser_state.gerrit_orig_line += 1
+        parser_state.gerrit_new_line += 1
+        # logging.debug('Unchanged start: ' + str(in_start))
```
Comment (dlatypov) [inline src/patch_parser.py]: reminder to drop these commented out logs.

**Review by dlatypov — CHANGES_REQUESTED** [review]
LGTM, just would want to clean it up a bit more.

**Comment by jaimehisao** [comment]
Code has been cleaned up!

**Review by dlatypov — APPROVED** [review]
LGTM, thanks!
